datafusion/expr/src/tree_node/expr.rs

```diff
@@ -0,0 +1,360 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Tree node implementation for logical expr
+
+use crate::expr::{
+    AggregateFunction, Between, BinaryExpr, Case, Cast, GetIndexedField, GroupingSet,
+    Like, Sort, TryCast, WindowFunction,
+};
+use crate::Expr;
+use datafusion_common::{tree_node::TreeNode, Result};
+
+impl TreeNode for Expr {
```

Comment:
There is some kind of duplication logic, I think maybe you can provide a default implementation for `map_children` in the  `TreeNode` trait.